### PR TITLE
Create option to maintain structure of your data.

### PR DESCRIPTION
**Pull request by Jatidude**
This is my first ever contribution to an open source repository so if i did something wrong just let me know. I changed the README to reflect the changes that I made but It could probable be cleaned up.

### CODE REVIEW

File: src/arrayToTree.ts [hunk 1]
```diff
@@ -35,11 +36,19 @@ export function arrayToTree (items: Item[], config: Config = { id: 'id', parentI
...
+        lookup[itemId] = { children: [] }
+      } else {
+        lookup[itemId] = { data: null, children: [] }
+      }
     }
 
     // add the current item's data to the item in the lookup table
-    lookup[itemId].data = item
+    if(config.maintainStructure) {
+      lookup[itemId] = {children: lookup[itemId].children, ...item}
```
Comment (philipstanislaus) [inline src/arrayToTree.ts]: Please spread the `item` first so that it cannot overwrite `children`.

File: src/arrayToTree.ts [hunk 2]
```diff
@@ -5,19 +5,20 @@ export interface Item {
 }
 
 export interface TreeItem {
-  data: Item | null
+  [key: string]: any,
```
Comment (philipstanislaus) [inline src/arrayToTree.ts]: Please keep `data` as an optional field for better typing: `data: Item | null`

File: README.md [hunk 3]
```diff
@@ -71,7 +72,7 @@ const tree = arrayToTree([
     { num: '1941', ref: '418', custom: 'de' },
     { num: '1', ref: '418', custom: 'ZZZz' },
     { num: '418', ref: null, custom: 'ü'},
-], { id: 'num', parentId: 'ref' })
+], { id: 'num', parentId: 'ref', maintainStructure: false })
```
Comment (philipstanislaus) [inline README.md]: Please remove `maintainStructure` here, it should be an optional field.

File: src/arrayToTree.ts [hunk 4]
```diff
@@ -5,19 +5,20 @@ export interface Item {
...
   children: TreeItem[]
 }
 
 export interface Config {
   id: string,
   parentId: string,
+  maintainStructure: boolean,
```
Comment (philipstanislaus) [inline src/arrayToTree.ts]: Could you mark all Config parameters as optional? `id?: string`, etc...

Comment (philipstanislaus) [reply src/arrayToTree.ts]: That should be done not just for maintainStructure, but parameter imo.

File: README.md [hunk 5]
```diff
@@ -61,6 +61,7 @@ You can provide a second argument to arrayToTree with configuration options. Rig
 
 - `id`: key of the id field of the item
 - `parentId`: key of the parent's id field of the item
+- `maintainStructure`: boolean to keep the current structure of your data. Not put it in a `data` field
```
Comment (philipstanislaus) [inline README.md]: Can we use a shorter descriptor? Maybe we make the `data` name adjustable with an option `{ dataField: "data" }`, and spread the properties into the top layer if `{ dataField: null }`?

**Comment by Jatidude** [comment]
@philipstanislaus is there something else I need to do before this gets merged? Sorry if I'm jumping the gun just want to make sure there's nothing more you need me to do.

**Comment by philipstanislaus** [comment]
@Jatidude sorry for the late reply, I was travelling. Just merged, will deploy in a bit!

**Comment by philipstanislaus** [comment]
@Jatidude all done, released as v1.2.0: https://www.npmjs.com/package/performant-array-to-tree

Thanks for your contribution!

**Pull request merged**